crates/phactory/src/prpc_service.rs

```diff
@@ -927,6 +969,155 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> Phactory<Platform>
+        )))?;
+        let key128 = derive_key_for_cluster_state(&key);
+        let nonce = rand::thread_rng().gen();
+        let dir = public_data_dir(&self.args.storage_path);
+        // Create directory if it does not exist
+        std::fs::create_dir_all(&dir).map_err(from_debug)?;
+        let fullname = dir.join(&filename);
+        let mut file = File::create(fullname).map_err(from_debug)?;
+        file.write_all(&system.ecdh_key.public())
+            .map_err(from_debug)?;
+
+        let mut enc_writer = aead::stream::new_aes128gcm_writer(key128, nonce, file);
+        enc_writer
+            .write_all(&CHECKPOINT_VERSION.to_be_bytes())
+            .map_err(from_debug)?;
+
+        serde_cbor::to_writer(enc_writer, &cluster_state).map_err(from_debug)?;
+        info!(%filename, "Cluster state saved");
+        Ok(pb::SaveClusterStateResponse {
+            block_number,
+            filename,
+        })
+    }
+
+    fn load_cluster_state(&mut self, filename: &str) -> RpcResult<()> {
+        let Some(system) = &self.system else {
+            return Err(from_display("System is uninitialized"));
+        };
+        if system.contract_cluster.is_some() {
+            return Err(from_display("Already in cluster"));
+        }
+        let Some(runtime_state) = &mut self.runtime_state else {
+            return Err(from_display("Runtime is uninitialized"));
+        };
+
+        let fullname = PathBuf::from(&self.args.storage_path).join(filename);
+        let mut reader = File::open(fullname).map_err(from_debug)?;
+        let mut pubkey: EcdhPublicKey = Default::default();
+        reader.read_exact(&mut pubkey).map_err(from_debug)?;
+
+        // Make sure the pubkey is not fake
+        if !runtime_state
+            .chain_storage
+            .is_worker_registered(&WorkerPublicKey(pubkey))
+        {
+            return Err(from_display(format!(
+                "The pubkey {} is not registered",
+                hex_fmt::HexFmt(&pubkey)
+            )));
+        }
+
+        let key = ecdh::agree(&system.ecdh_key, &pubkey).or(Err(from_display(
+            "Failed to derive ECDH key for cluster state",
+        )))?;
+        let key128 = derive_key_for_cluster_state(&key);
+        let mut dec_reader = aead::stream::new_aes128gcm_reader(key128, &mut reader);
+
+        let mut version = [0u8; 4];
+        dec_reader.read_exact(&mut version).map_err(from_debug)?;
+        let version = u32::from_be_bytes(version);
+        info!(version, "Loading cluster state");
+        if version != CHECKPOINT_VERSION {
+            return Err(from_display(format!(
+                "Incompatible cluster state version {version}, expected {CHECKPOINT_VERSION}",
+            )));
+        }
+        let recv_mq = &mut runtime_state.recv_mq;
+        let send_mq = &mut runtime_state.send_mq;
+        let cluster_state: ClusterState =
+            phala_mq::checkpoint_helper::using_dispatcher(recv_mq, move || {
```

Comment:
Why do you call `using_dispatcher()` here? It looks like `from_reader` doesn't access the mq.

Reply:
The deserializer of `mq::Receiver` would subscribe its topic with this dispatcher.

Reply:
Done